src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

```diff
@@ -589,12 +589,13 @@ private static bool CanBindToTheseConstructorParameters(ParameterInfo[] construc
 
             Debug.Assert(dictionary is not null);
 
-            Type dictionaryObjectType = dictionary.GetType();
+            Type? iDictionaryObjectType = FindOpenGenericInterface(typeof(IDictionary<,>), dictionary.GetType());
```

Comment:
Isn't this what `dictionaryType` already is? Do we need to do this lookup again?

Reply:
So what will happen in that case then? Seems like `iDictionaryObjectType` is going to be `null` if the object only implements `IReadOnlyDictionary`

Reply:
in case of IReadOnlyDictionary we always pass object of type Dictionary which still IDictionary<,> so we shouldn't get `null` for `iDictionaryObjectType`. That is why I used assert and not hard checking.

https://github.com/dotnet/runtime/blob/f79276122a60b78939d116af84cf6127420f752e/src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs#L535

Reply:
> in case of IReadOnlyDictionary we always pass object of type Dictionary which still IDictionary<,>

I don't believe that is true. Why can't I implement my own `IReadOnlyDictionary` that doesn't implement `IDictionary<,>`?

Reply:
Ah, I see what you are saying now. In the case that it could be `IReadOnlyDictionary`, we always instantiate a concrete `Dictionary<,>` and pass that object in to this method.

I think what would help the confusion here is to change this method to always use the `dictionaryType` parameter that is passed to the method. Add an assert to the top that it `dictionaryType.GetGenericTypeDefinition() == typeof(IDictionary<,>))`. And then fix up the call site here:

https://github.com/dotnet/runtime/blob/f79276122a60b78939d116af84cf6127420f752e/src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs#L540-L552

to not pass in `IReadOnlyDictionary`, but instead pass in `IDictionary<,>` to `BindConcreteDictionary`.

I think that would help in 2 ways:

1. it would make the code more readable/understandable to readers.
2. We wouldn't need to `FindOpenGenericInterface` twice when we are binding to a "normal" dictionary.

Reply:
`BindConcreteDictionary` is also called here with any potentially user-defined type that happens to implement `IDictionary<,>`:

https://github.com/dotnet/runtime/blob/d1938dca3fe8856c28850c9d8a8a1ce7d4066838/src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs#L364-L371

We should rename `BindConcreteDictionary` back to `BindDictionary`. I think it was named that way because it was originally intended to only be called from the call site you mention with plain-old concrete `Dictionary<,>` instances we created using `Activator.CreateInstance`, but that's apparently not the case.

Let's call it `BindDictionary` like we did in .NET 6 and get the setter (and now also the `TryGetValue` method) from the `dictionaryType` parameter.

https://github.com/dotnet/runtime/blob/943474ca16db7c65ba6cff4a89c3ebd219dde3e5/src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs#L446-L463

This means that we'll need to consistently pass in the reified `IDictionary<,>` (or `Dictionary<,>`) type. We already do that when we call it after `FindOpenGenericInterface` that I showed above. I think we just need to change `BindDictionaryInterface` to pass in `genericType` instead of `dictionaryType` as the second parameter to `BindConcreteDictionary`. Or `BindDictionary` after it's renamed.

Reply:
> I think we just need to change BindDictionaryInterface to pass in genericType (which is always Dictionary) instead of dictionaryType (which could be IReadOnlyDictionary) as the second parameter to BindConcreteDictionary/BindDictionary.

+1

That's a better idea than what I had above with "to not pass in `IReadOnlyDictionary`, but instead pass in `IDictionary<,>` to `BindConcreteDictionary`."